cdap-ranger/cdap-ranger-binding/src/main/java/co/cask/cdap/security/authorization/ranger/binding/RangerAuthorizer.java

```diff
@@ -0,0 +1,288 @@
+import co.cask.cdap.proto.id.ProgramId;
+import co.cask.cdap.proto.id.SecureKeyId;
+import co.cask.cdap.proto.id.StreamId;
+import co.cask.cdap.proto.security.Action;
+import co.cask.cdap.proto.security.Principal;
+import co.cask.cdap.proto.security.Privilege;
+import co.cask.cdap.proto.security.Role;
+import co.cask.cdap.security.spi.authorization.AbstractAuthorizer;
+import co.cask.cdap.security.spi.authorization.AuthorizationContext;
+import co.cask.cdap.security.spi.authorization.Authorizer;
+import co.cask.cdap.security.spi.authorization.UnauthorizedException;
+import org.apache.hadoop.security.UserGroupInformation;
+import org.apache.ranger.audit.provider.MiscUtil;
+import org.apache.ranger.plugin.audit.RangerDefaultAuditHandler;
+import org.apache.ranger.plugin.policyengine.RangerAccessRequestImpl;
+import org.apache.ranger.plugin.policyengine.RangerAccessResourceImpl;
+import org.apache.ranger.plugin.policyengine.RangerAccessResult;
+import org.apache.ranger.plugin.service.RangerBasePlugin;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.InetAddress;
+import java.util.Date;
+import java.util.Properties;
+import java.util.Set;
+
+/**
+ * This class implements {@link Authorizer} from CDAP and is responsible for interacting with Ranger to enforece
+ * authorization.
+ */
+public class RangerAuthorizer extends AbstractAuthorizer {
+  private static final Logger LOG = LoggerFactory.getLogger(RangerAuthorizer.class);
+
+  private static final String KEY_INSTANCE = "instance";
+  private static final String KEY_NAMESPACE = "namespace";
+  private static final String KEY_ARTIFACT = "artifact";
+  private static final String KEY_APPLICATION = "application";
+  private static final String KEY_DATASET = "dataset";
+  private static final String KEY_STREAM = "stream";
+  private static final String KEY_PROGRAM = "program";
+  private static final String KEY_DATASET_MODULE = "dataset_module";
+  private static final String KEY_DATASET_TYPE = "dataset_type";
+  private static final String KEY_SECUREKEY = "securekey";
+
+  private static final String RESOURCE_SEPARATOR = "#";
+
+
+  private static volatile RangerBasePlugin rangerPlugin = null;
+  private AuthorizationContext context;
+  private String instanceName;
+
+  @Override
+  public synchronized void initialize(AuthorizationContext context) throws Exception {
+    this.context = context;
+    Properties properties = context.getExtensionProperties();
+    instanceName = properties.containsKey(Constants.INSTANCE_NAME) ?
+      properties.getProperty(Constants.INSTANCE_NAME) : "cdap";
+    if (rangerPlugin == null) {
+      try {
+        UserGroupInformation ugi = UserGroupInformation.getLoginUser();
+        if (ugi != null) {
+          MiscUtil.setUGILoginUser(ugi, null);
+        }
+        LOG.debug("Initializing Ranger CDAP Plugin with UGI {}", ugi);
+      } catch (Throwable t) {
+        LOG.error("Error getting principal.", t);
```

Comment:
Good point. Although I think we need to decide as what's the right behavior for cdap when it fails to instantiate authorization plugin. Should cdap startup itself fail or should it start and then all the authorization will fail. First has the advantage that of failing early on although it limits the user in being able to start cdap successfully. The second will make cdap start although the most of the operations will fail since all the authorization will fail. Since enforce check for initialization the error the user will see will point to the cause.
@poornachandra @anwar6953 : Thoughts ?